File: src/shared_toolkit/ui/widgets/atomic/custom_line_edit.py
from PyQt6.QtCore import QRectF, Qt, QTimer
from PyQt6.QtGui import QColor, QPainter, QPainterPath, QPen, QRegion
from PyQt6.QtWidgets import QLineEdit

from shared_toolkit.ui.managers.theme_manager import ThemeManager

from ..helpers.underline_painter import UnderlineConfig, draw_bottom_underline

class CustomLineEdit(QLineEdit):
    RADIUS = 6

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setAttribute(Qt.WidgetAttribute.WA_StyledBackground, True)

        self.setProperty("custom-line-edit", True)

        self.setProperty("class", "primary")
        self.theme_manager = ThemeManager.get_instance()
        try:
            self.theme_manager.theme_changed.connect(self.update)
        except Exception:
            pass

        self.setStyleSheet("""
            QLineEdit {
                selection-background-color: transparent;
                background-color: transparent;
            }
        """)

    def _style_prefix(self) -> str:
        btn_class = str(self.property("class") or "")
        return "button.primary" if btn_class == "primary" else "button.default"

    def resizeEvent(self, e):

        super().resizeEvent(e)

    def focusInEvent(self, event):
        super().focusInEvent(event)

        QTimer.singleShot(0, self.update)

    def focusOutEvent(self, event):
        super().focusOutEvent(event)

        QTimer.singleShot(0, self.update)

    def paintEvent(self, e):

        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        r = self.rect()

        bg_color = self.theme_manager.get_color("dialog.input.background")
        painter.setBrush(bg_color)
        painter.setPen(Qt.PenStyle.NoPen)
        radius = self.RADIUS
        rr = QRectF(r).adjusted(0.5, 0.5, -0.5, -0.5)
        painter.drawRoundedRect(rr, radius, radius)
        painter.end()

        super().paintEvent(e)

        try:
            painter = QPainter(self)
            painter.setRenderHint(QPainter.RenderHint.Antialiasing)

            thin = QColor(self.theme_manager.get_color("input.border.thin"))
            alpha = max(8, int(thin.alpha() * 0.66))
            thin.setAlpha(alpha)
            pen = QPen(thin)
            pen.setWidthF(0.66)
            pen.setCapStyle(Qt.PenCapStyle.FlatCap)
            painter.setPen(pen)
            painter.setBrush(Qt.BrushStyle.NoBrush)
            painter.drawRoundedRect(rr, radius, radius)

            if self.hasFocus():
                underline_config = UnderlineConfig(
                    color=self.theme_manager.get_color("accent"),
                    alpha=255,
                    thickness=1.0
                )
            else:
                underline_config = UnderlineConfig(alpha=120, thickness=1.0)

            draw_bottom_underline(painter, r, self.theme_manager, underline_config)

            painter.end()
        except Exception:
            pass

File: cpp/toolkit/src/atomic/custom_line_edit.cpp
#include "sli/toolkit/atomic/custom_line_edit.h"

#include <QColor>
#include <QFocusEvent>
#include <QPainter>
#include <QPaintEvent>
#include <QPen>
#include <QRectF>
#include <Qt>

#include "sli/toolkit/theme.h"

namespace sli::toolkit {

CustomLineEdit::CustomLineEdit(QWidget* parent) : QLineEdit(parent) {
  setFrame(false);
  setTextMargins(kHPadding, kVPadding, kHPadding, kVPadding);
}

void CustomLineEdit::setUnderlineColor(const QColor& color) {
  underlineColor_ = color;
  update();
}

void CustomLineEdit::setFocusedUnderlineColor(const QColor& color) {
  focusedUnderlineColor_ = color;
  update();
}

void CustomLineEdit::setUnderlineThickness(double thickness) {
  underlineThickness_ = thickness;
  update();
}

void CustomLineEdit::paintEvent(QPaintEvent* event) {
  QPainter p(this);
  p.setRenderHint(QPainter::Antialiasing);
  const Palette& palette = Theme::palette();

  const QRectF r = QRectF(rect()).adjusted(0.5, 0.5, -0.5, -0.5);
  p.setPen(QPen(palette.border, 1));
  p.setBrush(palette.base);
  p.drawRoundedRect(r, kRadius, kRadius);

  // Optional underline accent.
  const QColor underline = hasFocus()
                                ? focusedUnderlineColor_.value_or(palette.accent)
                                : underlineColor_.value_or(QColor(0, 0, 0, 0));
  if (underline.alpha() > 0) {
    p.setPen(QPen(underline, underlineThickness_));
    const int y = r.bottom();
    p.drawLine(static_cast<int>(r.left() + kRadius), y,
               static_cast<int>(r.right() - kRadius), y);
  }
  p.end();
  QLineEdit::paintEvent(event);
}

void CustomLineEdit::focusInEvent(QFocusEvent* event) {
  QLineEdit::focusInEvent(event);
  update();
}

void CustomLineEdit::focusOutEvent(QFocusEvent* event) {
  QLineEdit::focusOutEvent(event);
  update();
}

}  // namespace sli::toolkit
